### PR TITLE
Complex number: soundness fixes, type promotion and caching

### PR DESCRIPTION
**Pull request by sergillam**
This PR implements the PR A quick-wins from the complex number improvement plan,
addressing correctness, soundness and performance issues without introducing new files
or structural refactoring.

### CODE REVIEW

File: src/python-frontend/function_call_expr.cpp [hunk 1]
```diff
@@ -3597,7 +3597,11 @@ function_call_expr::get_dispatch_table()
          exprt b = member_exprt(num, "imag", double_type());
          exprt c = member_exprt(den, "real", double_type());
          exprt d = member_exprt(den, "imag", double_type());
-
+         // P21: ZeroDivisionError guard — CPython raises when divisor == (0+0j).
+         exprt zero = from_double(0.0, double_type());
+         if (c == zero && d == zero)
+           return converter_.get_exception_handler().gen_exception_raise(
+             "ZeroDivisionError", "complex division by zero");
```
Comment (brcfarias) [inline src/python-frontend/function_call_expr.cpp]: Please, add a test case involving division by zero, including caught and uncaught exceptions.

**Review by brcfarias — CHANGES_REQUESTED** [review]

**Review by lucasccordeiro — APPROVED** [review]
Thanks for submitting this PR, @sergillam.
